### PR TITLE
Nazar/feature/out of stock viwe v1

### PR DESCRIPTION
**Pull request by Nazar-kob**

### CODE REVIEW

File: frontend/src/pages/Home/components/OutOfStock/OutOfStock.tsx [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import { useQuery } from "@tanstack/react-query";
+import { Spinner } from "../../../../components";
+import { API_BASE_URL } from "../../../../constants";
+import { UserProductsOut } from "../../../../main.type";
+import { GET_USER_PRODUCTS, TOKEN_KEY } from "../../../../services";
+import style from "./OutOfStock.module.css";
+import { Product } from "./Product";
+import { ResetBtn } from "./ResetBtn";
+
+type ProductId = Pick<UserProductsOut, "id">;
```
Comment (Taraslysak) [inline frontend/src/pages/Home/components/OutOfStock/OutOfStock.tsx:10]: What does this component do?

**Review by Taraslysak — APPROVED** [review]
Good job!

File: backend/app/router/product.py [hunk 2]
```diff
@@ -93,6 +93,24 @@ def update_product(
     return product
 
 
+@router.patch("/", status_code=status.HTTP_200_OK)
+def reset_product_out_of_stock(
+    business: m.Business = Depends(get_business_from_cur_user),
+    db: Session = Depends(get_db),
+):
+
+    log(log.INFO, "reset_product_out_of_stock")
```
[the remark this answers is not in the document]
Comment (Nazar-kob) [reply backend/app/router/product.py:102]: Thanks for advice I added)

File: frontend/src/pages/Home/components/OutOfStock/OutOfStock.tsx (hunk 1 above)
Comment (Nazar-kob) [reply frontend/src/pages/Home/components/OutOfStock/OutOfStock.tsx:10]: It resets all products

File: frontend/src/main.type.ts [hunk 3]
```diff
@@ -44,3 +44,19 @@ export type ManagerOutletContext = {
...
 };
+
+// this enum already exists and in future will replace in the project
+enum ItemUnit {
+  by_kilogram = "Kg",
+  by_unit = "Unit",
+  by_both = "by_both",
+}
+
+export type UserProductsOut = {
```
[the remark this answers is not in the document]
Comment (Nazar-kob) [reply frontend/src/main.type.ts:55]: I renamed that it will be clear

File: backend/app/schema/product_schemas.py [hunk 4]
```diff
@@ -36,10 +37,11 @@ class UpdateProduct(BaseModel):
     price: Optional[float]
     sold_by: Optional[m.SoldBy]
     image: Optional[str]
+    is_out_of_stock: Optional[bool]
 
     @root_validator
     def any_of(cls, v):
-        if not any(v.values()):
+        if len(v.values()) == sum([1 for value in v.values() if value is None]):
```
Comment (kipod) [inline backend/app/schema/product_schemas.py:44]: ```bool([i for i in l if i is not None])```
or
```[i for i in l if i is not None]```

File: frontend/src/pages/Home/components/OutOfStock/OutOfStock.tsx [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+      const res = await fetch(`${API_BASE_URL}/product/`, {
+        method: "GET",
+        headers: {
+          Authorization: `Bearer ${localStorage.getItem(TOKEN_KEY)}`,
+        },
+      });
+
+      if (!res.ok) {
+        localStorage.removeItem(TOKEN_KEY);
+        return false;
```
Comment (kipod) [inline frontend/src/pages/Home/components/OutOfStock/OutOfStock.tsx:25]: return empty array

File: frontend/src/pages/Home/components/OutOfStock/Product.tsx [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  price,
+  is_out_of_stock,
+}: Omit<TypeProductsOut, "sold_by">) => {
+  const mutationIsOutOfStock = useMutation({
+    mutationFn: isOutOfStock,
+    onSuccess: async (data) => {
+      queryClient.invalidateQueries([GET_USER_PRODUCTS]);
+    },
+    onError: async (err) => {
+      console.log(err, "mutationIsOutOfStock");
```
Comment (kipod) [inline frontend/src/pages/Home/components/OutOfStock/Product.tsx:21]: ```console.error```

File: frontend/src/pages/Home/components/OutOfStock/ResetBtn.tsx [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+const ResetBtn = () => {
+  const { openModal } = useOutletContext<ManagerOutletContext>();
+
+  const mutateResetOutOfStock = useMutation({
+    mutationFn: resetOutOfStock,
+    onSuccess: async () => {
+      queryClient.invalidateQueries([GET_USER_PRODUCTS]);
+    },
+    onError: async (err) => {
+      console.log(err, "mutateResetOutOfStock");
```
Comment (kipod) [inline frontend/src/pages/Home/components/OutOfStock/ResetBtn.tsx:22]: console.error()

**Review by kipod — APPROVED** [review]
Thank you

File: backend/app/schema/product_schemas.py (hunk 4 above)
Comment (Nazar-kob) [reply backend/app/schema/product_schemas.py]: Thanks

File: frontend/src/pages/Home/components/OutOfStock/OutOfStock.tsx (hunk 5 above)
Comment (Nazar-kob) [reply frontend/src/pages/Home/components/OutOfStock/OutOfStock.tsx]: I got you

File: frontend/src/pages/Home/components/OutOfStock/Product.tsx (hunk 6 above)
Comment (Nazar-kob) [reply frontend/src/pages/Home/components/OutOfStock/Product.tsx]: thanks

File: frontend/src/pages/Home/components/OutOfStock/ResetBtn.tsx (hunk 7 above)
Comment (Nazar-kob) [reply frontend/src/pages/Home/components/OutOfStock/ResetBtn.tsx]: done

**Pull request merged**